File: cole/day09_merge-two-sorted-lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        # initialize
        out, l = None, None
        if not list1 and not list2:
            return None
        elif list1 and list2:
            if list1.val < list2.val:
                out = list1
                list1 = list1.next
                l = out
            else:
                out = list2
                list2 = list2.next
                l = out
        elif list1:
            return list1
        else:
            return list2
        # merge
        while list1 and list2:
            if list1.val < list2.val:
                tmp = list1.next
                l.next = list1
                list1 = tmp
                l = l.next
            else:
                tmp = list2.next
                l.next = list2
                list2 = tmp
                l = l.next
        if list1:
            l.next = list1
        else:
            l.next = list2
        return out

File: cole/day09_merge-two-sorted-lists/merge-two-sorted-lists.c
// another 100%

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *mergeTwoLists(struct ListNode *list1, struct ListNode *list2)
{
    // obvious cases
    switch (((list2 != NULL) << 1) | (list1 != NULL))
    {
    case 0:
        return NULL;
    case 1:
        return list1;
    case 2:
        return list2;
    }
    // initialize output list
    struct ListNode *out, *l, *tmp;
    if (list1->val < list2->val)
    {
        out = list1;
        l = list1;
        list1 = list1->next;
    }
    else
    {
        out = list2;
        l = list2;
        list2 = list2->next;
    }
    // generate output in-place
    while (list1 && list2)
    {
        if (list1->val < list2->val)
        {
            tmp = list1->next;
            l->next = list1;
            l = l->next;
            list1 = tmp;
        }
        else
        {
            tmp = list2->next;
            l->next = list2;
            l = l->next;
            list2 = tmp;
        }
    }
    l->next = list1 ? list1 : list2;
    return out;
}